### PR TITLE
Adds EventDispatcher and change event to DeviceOrientationControls and FlyControls

### PR DESCRIPTION
**Pull request by gsimone**
Implementation is lifted from OrbitControls, it checks dot product for quaternion change and dist Squared for position change

Types now extend EventDispatcher

### CODE REVIEW

File: examples/jsm/controls/DeviceOrientationControls.js [hunk 1]
```diff
@@ -107,28 +106,40 @@ var DeviceOrientationControls = function ( object ) {
...
-			setObjectQuaternion( scope.object.quaternion, alpha, beta, gamma, orient );
+				var gamma = device.gamma ? MathUtils.degToRad( device.gamma ) : 0; // Y''
 
-		}
+				var orient = scope.screenOrientation ? MathUtils.degToRad( scope.screenOrientation ) : 0; // O
 
+				setObjectQuaternion( scope.object.quaternion, alpha, beta, gamma, orient );
 
-	};
+				if ( 8 * ( 1 - lastQuaternion.dot( scope.object.quaternion ) ) > EPS ) {
```
Comment (Mugen87) [inline examples/jsm/controls/DeviceOrientationControls.js:131]: May I ask: Where does this formula come from? Meaning the usage of `8` and `0.000001` as epsilon value.

Comment (gsimone) [reply examples/jsm/controls/DeviceOrientationControls.js:131]: OrbitControls

**Comment by Mugen87** [comment]
Can you please editr the files in the `examples/js` directory instead and then generate the modules via:

```
node utils/modularize.js
```

File: examples/jsm/controls/DeviceOrientationControls.js [hunk 2]
```diff
@@ -107,28 +106,40 @@ var DeviceOrientationControls = function ( object ) {
...
 	};
 
-	this.update = function () {
+	this.update = ( function () {
 
-		if ( scope.enabled === false ) return;
+		const lastQuaternion = new Quaternion();
 
-		var device = scope.deviceOrientation;
+		return () => {
```
Comment (Mugen87) [inline examples/jsm/controls/DeviceOrientationControls.js:113]: Please use no arrow functions.

File: examples/jsm/controls/FlyControls.js [hunk 3]
```diff
@@ -186,23 +187,40 @@ var FlyControls = function ( object, domElement ) {
...
+			this.object.quaternion.multiply( this.tmpQuaternion );
 
-	};
+			// expose the rotation vector for convenience
+			this.object.rotation.setFromQuaternion( this.object.quaternion, this.object.rotation.order );
+
+			if (
+				lastPosition.distanceToSquared( this.object.position ) > EPS ||
+				8 * ( 1 - lastQuaternion.dot( this.object.quaternion ) ) > EPS
+			) {
```
Comment (Mugen87) [inline examples/jsm/controls/FlyControls.js:213]: `else` missing.

File: examples/jsm/controls/DeviceOrientationControls.js (hunk 2 above)
Comment (gsimone) [reply examples/jsm/controls/DeviceOrientationControls.js]: Done

**Comment by gsimone** [comment]
Made the changes to the js versions and built the jsm ones 👍

**Comment by Mugen87** [comment]
Looks good now!

Can you please also update the docs for both classes? The idea is to add a new section called `Events` similar to the other pages e.g. https://threejs.org/docs/index.html#examples/en/controls/OrbitControls

**Comment by gsimone** [comment]
I only added the Change event, do you think it's worth adding start and end? ( taking Orbit as an example ) Should be easy enough to do.

**Comment by Mugen87** [comment]
It's probably good if the controls fire similar events if possible. Would be great if you can have a look! 👍

**Comment by gsimone** [comment]
I was adding the events but realized that while OrbitControls have precise start and end times ( ie. when the user clicks ), fly and deviceOrientation don't, it's always being updated and doesn't really start/end

So I just committed the doc changes, if you think the two controls need more events I can look into that but I would open a new issue and PR

**Review by Mugen87 — APPROVED** [review]

**Comment by gsimone** [comment]
Hi 👋
Is there anything missing to merge this?

**Comment by Mugen87** [comment]
> Is there anything missing to merge this?

@mrdoob still needs to review the PR 👍

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**